File: backend/app/geometry/coordinate_transform.py
"""
Pure, framework-free coordinate transform functions implementing the
project <-> local engineering half of the chain defined in
docs/architecture/coordinate-strategy.md. This is a deliberate port of
frontend/src/geometry/coordinateTransform.ts's project_to_local /
local_to_project -- same derivation, same sign convention, verified by the
same hand-computed orientation test cases (see
tests/geometry/test_coordinate_transform.py) so a bearing-sign bug can't
silently diverge between the two language implementations.

The backend never needs the local <-> viewer stage (that is a Three.js
rendering-layer concept, frontend-only).
"""

from __future__ import annotations

import math

import numpy as np
import numpy.typing as npt

from app.domain.coordinates import LocalCoordinate, LocalFrameDefinition, ProjectCoordinate


def _local_basis(line_bearing_radians: float) -> tuple[float, float]:
    return math.cos(line_bearing_radians), math.sin(line_bearing_radians)


def project_to_local(point: ProjectCoordinate, frame: LocalFrameDefinition) -> LocalCoordinate:
    d_easting = point.easting - frame.mast_centre_project.easting
    d_northing = point.northing - frame.mast_centre_project.northing
    cos_b, sin_b = _local_basis(frame.line_bearing_radians)

    x = d_easting * cos_b - d_northing * sin_b
    y = d_easting * sin_b + d_northing * cos_b
    z = point.elevation - frame.mast_centre_project.elevation

    return LocalCoordinate(x=x, y=y, z=z)


def local_to_project(point: LocalCoordinate, frame: LocalFrameDefinition) -> ProjectCoordinate:
    cos_b, sin_b = _local_basis(frame.line_bearing_radians)

    d_easting = point.x * cos_b + point.y * sin_b
    d_northing = -point.x * sin_b + point.y * cos_b

    return ProjectCoordinate(
        easting=frame.mast_centre_project.easting + d_easting,
        northing=frame.mast_centre_project.northing + d_northing,
        elevation=frame.mast_centre_project.elevation + point.z,
    )


def project_to_local_array(
    easting: npt.NDArray[np.floating],
    northing: npt.NDArray[np.floating],
    elevation: npt.NDArray[np.floating],
    frame: LocalFrameDefinition,
) -> tuple[npt.NDArray[np.floating], npt.NDArray[np.floating], npt.NDArray[np.floating]]:
    """
    Vectorized form of project_to_local for converting a whole clipped
    point array at once (used by processing/las_clip.py). Shares
    `_local_basis` with the scalar function above so the two can't
    silently diverge -- see test_coordinate_transform.py's
    test_array_form_matches_scalar_form_elementwise.
    """
    cos_b, sin_b = _local_basis(frame.line_bearing_radians)
    d_easting = easting - frame.mast_centre_project.easting
    d_northing = northing - frame.mast_centre_project.northing

    x = d_easting * cos_b - d_northing * sin_b
    y = d_easting * sin_b + d_northing * cos_b
    z = elevation - frame.mast_centre_project.elevation

    return x, y, z

File: frontend/src/geometry/coordinateTransform.ts
/**
 * Pure, framework-free coordinate transform functions implementing the chain
 * defined in docs/architecture/coordinate-strategy.md:
 *
 *   project <-> local engineering <-> viewer
 *
 * These functions are the only place the bearing sign convention and axis
 * derivation are resolved. Do not re-derive coordinate math elsewhere.
 */

import type {
  LocalCoordinate,
  LocalFrameDefinition,
  ProjectCoordinate,
  ViewerCoordinate,
  ViewerFrameDefinition,
} from "../domain/coordinates";
import { localCoordinate, projectCoordinate, viewerCoordinate } from "../domain/coordinates";

/**
 * Local +Y (longitudinal) is defined to point along the line bearing,
 * measured clockwise from project/grid north. Local +X (transverse) and
 * local +Y must form a right-handed system with +Z up (X x Y = Z).
 *
 * With bearing beta clockwise from north, the longitudinal direction in
 * (Easting, Northing) is (sin(beta), cos(beta)). Solving X . Y = 0,
 * |X| = 1 and X x Y = Z for a horizontal X gives X = (cos(beta), -sin(beta))
 * as the unique right-handed solution (verified: cos(beta)*cos(beta) -
 * (-sin(beta))*sin(beta) = 1). Expressing a project-space offset in this
 * orthonormal basis is equivalent to a standard CCW rotation of the offset
 * vector by beta.
 */
function localBasis(lineBearingRadians: number): {
  cos: number;
  sin: number;
} {
  return { cos: Math.cos(lineBearingRadians), sin: Math.sin(lineBearingRadians) };
}

export function projectToLocal(
  point: ProjectCoordinate,
  frame: LocalFrameDefinition
): LocalCoordinate {
  const dEasting = point.easting - frame.mastCentreProject.easting;
  const dNorthing = point.northing - frame.mastCentreProject.northing;
  const { cos, sin } = localBasis(frame.lineBearingRadians);

  const x = dEasting * cos - dNorthing * sin;
  const y = dEasting * sin + dNorthing * cos;
  const z = point.elevation - frame.mastCentreProject.elevation;

  return localCoordinate(x, y, z);
}

export function localToProject(
  point: LocalCoordinate,
  frame: LocalFrameDefinition
): ProjectCoordinate {
  const { cos, sin } = localBasis(frame.lineBearingRadians);

  const dEasting = point.x * cos + point.y * sin;
  const dNorthing = -point.x * sin + point.y * cos;

  return projectCoordinate(
    frame.mastCentreProject.easting + dEasting,
    frame.mastCentreProject.northing + dNorthing,
    frame.mastCentreProject.elevation + point.z
  );
}

export function localToViewer(
  point: LocalCoordinate,
  frame: ViewerFrameDefinition
): ViewerCoordinate {
  return viewerCoordinate(
    point.x - frame.renderOriginLocal.x,
    point.y - frame.renderOriginLocal.y,
    point.z - frame.renderOriginLocal.z
  );
}

export function viewerToLocal(
  point: ViewerCoordinate,
  frame: ViewerFrameDefinition
): LocalCoordinate {
  return localCoordinate(
    point.x + frame.renderOriginLocal.x,
    point.y + frame.renderOriginLocal.y,
    point.z + frame.renderOriginLocal.z
  );
}

export function projectToViewer(
  point: ProjectCoordinate,
  localFrame: LocalFrameDefinition,
  viewerFrame: ViewerFrameDefinition
): ViewerCoordinate {
  return localToViewer(projectToLocal(point, localFrame), viewerFrame);
}

export function viewerToProject(
  point: ViewerCoordinate,
  localFrame: LocalFrameDefinition,
  viewerFrame: ViewerFrameDefinition
): ProjectCoordinate {
  return localToProject(viewerToLocal(point, viewerFrame), localFrame);
}
